Unexpected invocation of NewXxxKindREST() in controller during package v1.init()

I've spotted that when defining a custom REST storage, its constructor is being called both from apiserver and controller-manager binaries (panic() is inserted manually in the code):
```
$ bin/controller-manager --kubeconfig=kubeconfig
panic: panic

goroutine 1 [running]:
test/pkg/handler.NewHandler(0xc42078a380, 0xc42068f6b0, 0xc42078a380, 0xc42068f6b0, 0x0)
        /go/src/test/pkg/handler/handler.go:56 +0x99
test/pkg/apis/apps/v1.NewApplicationREST(0x3c95540, 0xc4207fe640)
        /go/src/test/pkg/apis/apps/v1/application_rest.go:203 +0xc0
test/pkg/apis/apps/v1.init()
        /go/src/test/pkg/apis/apps/v1/zz_generated.api.register.go:49 +0x1eb
test/pkg/controller/application.init()
        /go/src/test/pkg/controller/application/zz_generated.api.register.go:99 +0x5d
test/pkg/controller.init()
        /go/src/test/pkg/controller/zz_generated.api.register.go:41 +0x44
main.init()
        /go/src/test/cmd/controller/main.go:31 +0x5a```
```

This happens because NewXxxRest() is called during package-level variable initialization in v1/zz_generated.api.register.go:

```
        appsApplicationStorage = builders.NewApiResourceWithStorage( // Resource status endpoint
                apps.InternalApplication,
                ApplicationSchemeFns{},
                func() runtime.Object { return &Application{} },     // Register versioned resource
                func() runtime.Object { return &ApplicationList{} }, // Register versioned resource list
                NewApplicationREST(),
        )
```

NewXxxREST() constructor may contain some other logic besides simply initializing the custom REST registry type. For example it may establish a connection to some service. And we'll end up with an extra unused connection to the external service, established from the controller manager.

I think we should either fix this or explicitly state (in the documentation and/or in the generated code) that NewXxxREST() should only construct new empty instance of XxxREST without performing any heavy initialization.

--- a/cmd/apiregister-gen/generators/versioned_generator.go
+++ b/cmd/apiregister-gen/generators/versioned_generator.go
@@ -39,14 +39,28 @@ func CreateVersionedGenerator(apiversion *APIVersion, apigroup *APIGroup, filena
 	}
 }
 
+func hasSubresources(version *APIVersion) bool {
+	for _, v := range version.Resources {
+		if len(v.Subresources) != 0 {
+			return true
+		}
+	}
+	return false
+}
+
 func (d *versionedGenerator) Imports(c *generator.Context) []string {
-	return []string{
+	imports := []string{
 		"metav1 \"k8s.io/apimachinery/pkg/apis/meta/v1\"",
 		"k8s.io/apimachinery/pkg/runtime",
 		"github.com/kubernetes-incubator/apiserver-builder/pkg/builders",
 		"k8s.io/apimachinery/pkg/runtime/schema",
 		d.apigroup.Pkg.Path,
 	}
+	if hasSubresources(d.apiversion) {
+		imports = append(imports, "k8s.io/apiserver/pkg/registry/rest")
+	}
+
+	return imports
 }
 
 func (d *versionedGenerator) Finalize(context *generator.Context, w io.Writer) error {
@@ -64,7 +78,7 @@ var (
 			{{.Kind}}SchemeFns{},
 			func() runtime.Object { return &{{ $api.Kind }}{} },     // Register versioned resource
 			func() runtime.Object { return &{{ $api.Kind }}List{} }, // Register versioned resource list
-			New{{ $api.REST }}(),
+			New{{ $api.REST }},
 		)
 	{{ else -}}
 		{{$api.Group}}{{$api.Kind}}Storage = builders.NewApiResource( // Resource status endpoint
@@ -95,7 +109,7 @@ var (
 			builders.SchemeFnsSingleton,
 			func() runtime.Object { return &{{ $subresource.Request }}{} }, // Register versioned resource
 			nil,
-			&{{ $subresource.REST }}{ {{$api.Group}}.New{{$api.Kind}}Registry({{$api.Group}}{{$api.Kind}}Storage) },
+			func() rest.Storage { return &{{ $subresource.REST }}{ {{$api.Group}}.New{{$api.Kind}}Registry({{$api.Group}}{{$api.Kind}}Storage) } },
 		),
 		{{ end -}}
 		{{ end -}}

--- a/pkg/builders/api_versioned_resource_builder.go
+++ b/pkg/builders/api_versioned_resource_builder.go
@@ -61,15 +61,15 @@ func NewApiResourceWithStorage(
 	unversionedBuilder UnversionedResourceBuilder,
 	schemeFns SchemeFns,
 	new, newList func() runtime.Object,
-	storage rest.Storage) *versionedResourceBuilder {
+	RESTFunc func() rest.Storage) *versionedResourceBuilder {
 	v := &versionedResourceBuilder{
-		unversionedBuilder, schemeFns, new, newList, nil, storage, nil,
+		unversionedBuilder, schemeFns, new, newList, nil, RESTFunc, nil,
 	}
 	if new == nil {
 		panic(fmt.Errorf("Cannot call NewApiResourceWithStorage with nil new function."))
 	}
-	if storage == nil {
-		panic(fmt.Errorf("Cannot call NewApiResourceWithStorage with nil new storage."))
+	if RESTFunc == nil {
+		panic(fmt.Errorf("Cannot call NewApiResourceWithStorage with nil RESTFunc function."))
 	}
 	return v
 }
@@ -84,11 +84,11 @@ type versionedResourceBuilder struct {
 	// NewListFunc returns and empty unversioned instance of a resource List
 	NewListFunc func() runtime.Object
 
-	// Store is used to modify the default storage, mutually exclusive with RESTFunc
+	// StorageBuilder is used to modify the default storage, mutually exclusive with RESTFunc
 	StorageBuilder StorageBuilder
 
-	// REST a rest.Store implementation, mutually exclusive with StoreFunc
-	REST rest.Storage
+	// RESTFunc returns a rest.Storage implementation, mutually exclusive with StorageBuilder
+	RESTFunc func() rest.Storage
 
 	Storage rest.StandardStorage
 }
@@ -173,9 +173,9 @@ func (b *versionedResourceBuilder) registerEndpoints(
 		path = b.Unversioned.GetName()
 	}
 
-	if b.REST != nil {
+	if b.RESTFunc != nil {
 		// Use the REST implementation directly.
-		registry[path] = b.REST
+		registry[path] = b.RESTFunc()
 	} else {
 		// Create a new REST implementation wired to storage.
 		registry[path] = b.

